inttest/ap-ha3x3/ha3x3_test.go

```diff
@@ -127,64 +129,159 @@ spec:
                   - worker2
 `
 
-	manifestFile := "/tmp/happy.yaml"
-	s.PutFileTemplate(s.ControllerNode(0), manifestFile, planTemplate, nil)
+	ctx := s.Context()
 
-	out, err := s.RunCommandController(0, fmt.Sprintf("/usr/local/bin/k0s kubectl apply -f %s", manifestFile))
-	s.T().Logf("kubectl apply output: '%s'", out)
+	sshController, err := s.SSH(ctx, s.ControllerNode(0))
 	s.Require().NoError(err)
+	defer sshController.Disconnect()
+
+	if version, err := s.GetK0sVersion(s.ControllerNode(0)); s.NoError(err, "Failed to get the base k0s version") {
+		hasOldStack := version != s.k0sUpdateVersion && (strings.HasPrefix(version, "v1.27.") || strings.HasPrefix(version, "v1.28."))
+		s.T().Logf("Base k0s version: %q, has old stack: %v", version, hasOldStack)
+		s.checkKubeletConfigStackResources(ctx, sshController, hasOldStack)
+	}
 
-	ssh, err := s.SSH(s.Context(), s.WorkerNode(0))
+	sshWorker, err := s.SSH(ctx, s.WorkerNode(0))
 	s.Require().NoError(err)
-	defer ssh.Disconnect()
-	out, err = ssh.ExecWithOutput(s.Context(), "/var/lib/k0s/bin/iptables-save -V")
+	defer sshWorker.Disconnect()
+
+	iptablesModeBeforeUpdate, err := getIPTablesMode(ctx, sshWorker)
+	if !s.NoError(err) {
+		iptablesModeBeforeUpdate = ""
+	}
+
+	var createPlanOutput bytes.Buffer
+	err = sshController.Exec(ctx, "k0s kc create -f -", common.SSHStreams{
+		In:  strings.NewReader(planTemplate),
+		Out: &createPlanOutput,
+	})
 	s.Require().NoError(err)
-	iptablesVersionParts := strings.Split(out, " ")
-	iptablesModeBeforeUpdate := iptablesVersionParts[len(iptablesVersionParts)-1]
+	s.T().Log(strings.TrimSpace(createPlanOutput.String()))
 
 	client, err := s.AutopilotClient(s.ControllerNode(0))
 	s.Require().NoError(err)
-	s.NotEmpty(client)
 
 	// The plan has enough information to perform a successful update of k0s, so wait for it.
-	plan, err := aptest.WaitForPlanState(s.Context(), client, apconst.AutopilotName, appc.PlanCompleted)
+	s.T().Log("Waiting for autopilot plan to complete")
+	plan, err := aptest.WaitForPlanState(ctx, client, apconst.AutopilotName, appc.PlanCompleted)
 	s.Require().NoError(err)
+	s.T().Log("Autopilot plan completed")
 
 	// Ensure all state/status are completed
-	s.Equal(1, len(plan.Status.Commands))
-	cmd := plan.Status.Commands[0]
+	if s.Len(plan.Status.Commands, 1) {
+		cmd := plan.Status.Commands[0]
+		s.Equal(appc.PlanCompleted, cmd.State)
+		s.Equal(appc.PlanCompleted, cmd.State)
```

Comment:
This line is duplicated, it doesn't have any real impact so it's a nit, but let's remove it anyway.

Reply:
Killed it with :fire: